File: java/deepspace/src/deepspace/SpecificDamage.java
package deepspace;
import java.util.ArrayList;

public class SpecificDamage extends Damage{

    private ArrayList<WeaponType> weapons;
    
    SpecificDamage(ArrayList<WeaponType> wl, int s) {
	super(s);
	weapons=wl; 
    }

    //dada una lista devuelve el índice del tipo pasado
    //si no se encuntra devuelve -1
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t) {
	int salida=-1; //valor de salida erronea
	
	for(int i=0;i<w.size() && salida <0;i++) {
            if(w.get(i).getType() == t) {
                salida=i;
            }
        }
        return salida;
        
    }

    
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
	ArrayList<WeaponType> newWeapons=new ArrayList<>(); 
        ArrayList<Weapon> auxWeapons = new ArrayList<>(w);
	for(WeaponType wtype : weapons) {
	    int indice=arrayContainsType(auxWeapons,wtype);
	    if(indice >= 0) {
                newWeapons.add(wtype);
		auxWeapons.remove(indice); 
	    }
	}

	return new SpecificDamage(newWeapons, adjustShields(s));
        
    }

    //Se quiere descartar un tipo de nuestro daño pendiente
    //Incompatibilidad de tipos
    @Override
    public void discardWeapon(Weapon w) {
	for(WeaponType wt:weapons) {
            if(wt == w.getType()) {
                weapons.remove(wt);
                return;
            }
        }
    }

    @Override
    public boolean hasNoEffect() {
	return super.hasNoEffect() && weapons.isEmpty(); 
    }

    public ArrayList<WeaponType> getWeapons() {
	return weapons; 
    }

    @Override
    public String toString() {
	String repr="\n__Specific Damage__\n"
	    + super.toString()
	    + "Weapons: \n";

	if(weapons != null) for(WeaponType weapon : weapons)
	    repr += weapon;
	repr += "\n";

	return repr; 
	
    }
    
    @Override
    public SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
}

File: ruby/SpecificDamage.rb
# coding: utf-8

require_relative 'Damage'
require_relative 'SpecificDamageToUI'

module Deepspace

  class SpecificDamage < Damage

    def initialize(wl,s)
      super(s)
      @weapons = wl
    end

    public_class_method :new

    attr_reader :weapons

    def self.newCopy(d)
      SpecificDamage.new(d.weapons,d.nShields)
    end

    def getUIversion
      SpecificDamageToUI.new(self)
    end

    def arrayContainsType w, t
      indice= -1

      indice_aux = 0
      len = w.length

      while indice == -1 and indice_aux < len
        if w[indice_aux].type == t
          indice = indice_aux
        else
          indice_aux += 1
        end #if
      end #while

      indice
    end
   private :arrayContainsType

   def adjust(w,s)
     newNShields = adjustShields(s)

     auxWeap = w.clone
     newWeapons = Array.new
     @weapons.each { |weapon|
       ind = arrayContainsType(auxWeap,weapon)
       if(ind != -1)
         newWeapons << weapon
         auxWeap.delete_at(ind)
       end
     }

     return SpecificDamage.new(newWeapons,newNShields)
   end

   def discardWeapon(w)
     @weapons.size.times do |i|
       if(@weapons[i]==w.type)
         @weapons.delete_at(i)
         return
       end
     end
   end

   def hasNoEffect
     super && @weapons.size == 0
   end

   def to_s
     s="Weapons:\n"
     weapons.each do |w| s += w.to_s end
     s+= "\n"
     s+=super
     return s
   end

 end # SpecificDamage

end #module
